### PR TITLE
Cxx unit testing 471

### LINKED ISSUE #471: Setup C++ unit testing framework
This is going to be especially useful for some independent classes such as the CommandCenter being developed in the science design branch. Any unit test framework is on the table provided it works. No requirements to put it on Travis on this issue.

### PR DESCRIPTION
**Pull request by steamypotato**
# Assignee Section
 ## Description 
Unit testing for C++ code. This PR will include a basic example test directory in science just for testing purposes. Check the wiki 
 (/wiki/CXX-----Unit-Testing) for a detailed explanation on how everything works. 

### Steps for Testing 
1. `mkdir build && cd build` (in robot/rover) 
2. `cmake ..` (Make sure the installation of AUnit and UnixHostDuino are logged) 
3. `make` (We are looking for the 'ScienceTests' target to be built) 
4. `cd ../science/tests` 
5. `make`
6. `./ScienceTests.out` (Check that tests run)
7. `make clean` 

closes #471 

# Reviewer Section Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process. 
- [ ] Local Test Performed Successfully
- [ ] [General Integration Test (https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit)Performed Successfully For Pull Requests that do not include code changes, it is not required to perform the tests above.

### CODE REVIEW

**Comment by VimFreeman** [comment]
screenshot
Is this expected behaviour?

**Review by VimFreeman — COMMENTED** [review]
screenshot_1
Step 5 seems to fail.

File: robot/rover/CMakeLists.txt [hunk 1]
```diff
@@ -186,13 +202,12 @@ foreach(path ${INOFILES})
  if(${name} STREQUAL "Arm")
  set(PROJECT_LIBRARIES  Servo)
 
- # this particular project is important and should be supported, issue #409 opened for this
+     # this particular project is important and should be supported, issue #409 opened for this
```
Comment (MartensCedric) [inline robot/rover/CMakeLists.txt:205]: bad indent

Comment (MartensCedric) [inline robot/rover/CMakeLists.txt:205]: bad indent

File: robot/rover/CMakeLists.txt [hunk 2]
```diff
@@ -1,9 +1,9 @@
-cmake_minimum_required(VERSION 3.5.0)
+cmake_minimum_required(VERSION 3.10.2)
 
 set(CMAKE_SYSTEM_PROCESSOR arm)
 set(MCU mk66fx1m0)
 
-include(ExternalProject)
+#include(ExternalProject)
```
Comment (MartensCedric) [inline robot/rover/CMakeLists.txt:6]: why is this commented? Can we remove it?

File: robot/rover/CMakeTemplate.txt.in [hunk 3]
```diff
@@ -30,30 +30,34 @@ target_include_directories(${PROJECT_NAME}_EXE PUBLIC ${CMAKE_CURRENT_SOURCE_DIR
...
+        POST_BUILD
+        COMMAND ${ARM_GCC_OBJCOPY} -O ihex -R.eeprom ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${PROJECT_NAME}.elf ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${PROJECT_NAME}.hex
 )
 add_custom_command(OUTPUT ${PROJECT_NAME}_upload_command
-COMMAND ${TEENSY_LOADER_CLI} --mcu=${MCU} -s -v ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${PROJECT_NAME}.hex 
-DEPENDS ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${PROJECT_NAME}.hex 
-)
+        COMMAND ${TEENSY_LOADER_CLI} --mcu=${MCU} -s -v ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${PROJECT_NAME}.hex
+        DEPENDS ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${PROJECT_NAME}.hex
+        )
```
Comment (MartensCedric) [inline robot/rover/CMakeTemplate.txt.in:55]: you def wanna fix the indentation here

File: robot/rover/science/raman.ino [hunk 4]
```diff
@@ -6,10 +6,9 @@
 #include <include/commands/CommandCenter.h>
 #include <include/commands/ScienceCommandCenter.h>
 #include "include/Carousel.h"
-#include <cstdint>
 
-const uint8_t NUMBER_OF_STOPPABLES = 5;
-const uint8_t NUMBER_OF_UPDATABLES = 5;
+const int NUMBER_OF_STOPPABLES = 5;
+const int NUMBER_OF_UPDATABLES = 5;
```
Comment (MartensCedric) [inline robot/rover/science/raman.ino:11]: any reason for this? We're planning on using strong types everywhere, especially for comms

**Review by MartensCedric — CHANGES_REQUESTED** [review]

File: robot/rover/science/raman.ino (hunk 4 above)
Comment (steamypotato) [reply robot/rover/science/raman.ino]: (bad commit somehow made it to the PR)

**Comment by steamypotato** [comment]
> screenshot
> Is this expected behaviour?

yup i've had this bug for a while, it happens when you cross compile in cmake, its just a visual bug it doesn't affect anything although I havent found a solution for it yet

**Comment by VimFreeman** [comment]
Just tested again, this time on step 6, I do not have a file called ScienceTests.out in the science/tests directory

**Comment by MartensCedric** [comment]
@steamypotato could you fix this because Im gonna make unit tests

**Review by VimFreeman — DISMISSED** [review]
Just tested again and it works 👍

**Comment by MartensCedric** [comment]
you have conflicts but you're done

**Comment by MartensCedric** [comment]
@steamypotato the tests fail, could you 1) make it so the tests dont fail, 2) make sure that if travis fails, github CI also fails (something is clearly wrong in github CI if only travis fails!)

**Review by MartensCedric — APPROVED** [review]

**Comment by MartensCedric** [comment]
tim orally approves

**Pull request merged**